experimental/rank1_bnns/refining.py

```diff
@@ -0,0 +1,119 @@
+  posterior_var = tf.math.pow(posterior_scale, 2)
+  auxiliary_var = tf.math.pow(auxiliary_scale, 2)
+  conditional_mean = (prior_mean + (auxiliary_sample * posterior_var *
+                                    prior_var + (posterior_mean - prior_mean)
+                                    * (prior_var - auxiliary_var)
+                                    * prior_var)
+                      / (posterior_var * auxiliary_var + prior_var *
+                         (prior_var - auxiliary_var)))
+  conditional_var = posterior_var * prior_var * (prior_var - auxiliary_var) / \
+                    (auxiliary_var * posterior_var + prior_var *
+                     (prior_var - auxiliary_var))
+  return conditional_mean, tf.sqrt(conditional_var)
+
+
+def sample_rank1_auxiliaries(model, auxiliary_var_ratio):
+  for layer in model.layers:
+    if isinstance(layer, DenseRank1) or isinstance(layer, Conv2DRank1):
+      for rv_name, rv, regularizer in [('alpha', layer.alpha,
+                                        layer.alpha_regularizer),
+                                       ('gamma', layer.gamma,
+                                        layer.gamma_regularizer)]:
+        posterior_mean = rv.distribution.distribution.loc
+        posterior_scale = rv.distribution.distribution.scale
+        prior_mean = regularizer.mean
+        prior_scale = regularizer.stddev
+        auxiliary_scale_ratio = np.sqrt(auxiliary_var_ratio)
+        auxiliary_scale = tf.cast(auxiliary_scale_ratio * prior_scale,
+                                  dtype=posterior_mean.dtype)
+        a_mean, a_scale = get_auxiliary_posterior(posterior_mean,
+                                                  posterior_scale,
+                                                  prior_mean,
+                                                  prior_scale,
+                                                  auxiliary_scale)
+        auxiliary_sample = tfp.distributions.Normal(loc=a_mean,
+                                                    scale=a_scale).sample()
+        new_posterior_mean, new_posterior_scale = get_conditional_posterior(
+            posterior_mean,
+            posterior_scale,
+            prior_mean,
+            prior_scale,
+            auxiliary_scale,
+            auxiliary_sample)
+        new_prior_mean, new_prior_scale = get_conditional_prior(
+            prior_mean,
+            prior_scale,
+            auxiliary_scale,
+            auxiliary_sample)
+        for v in layer.variables:
+          if rv_name + '/mean' in v.name:
+            posterior_mean_variable = v
+            posterior_mean_variable.assign(new_posterior_mean)
+          if rv_name + '/stddev' in v.name:
+            posterior_scale_untransformed = v
+            posterior_scale_untransformed.assign(
+                tfp.math.softplus_inverse(new_posterior_scale))
+          regularizer.mean = new_prior_mean.numpy()
+          regularizer.stddev = new_prior_scale.numpy()
```

Comment:
unindent

Reply:
I think my editor was set to use 4 spaces. Fixed now!